### PR TITLE
Enable preserveOutputs for dynamic outputs and introduce preserveOutputWhitelist

### PR DESCRIPTION
**Pull request by semihokur**
- The changes enable preserveOutputs for dynamic outputs. 
- We also introduce 'preserveOutputWhitelist'. If the list is not empty, BuildXL will preserve all outputs. Otherwise, only given paths will be preserved. Devs can only whitelist the declared output paths which are either static file outputs or dynamic directories. They cannot specify the individual dynamic file outputs under directory outputs.

### CODE REVIEW

**Comment by semihokur** [comment]
I was not expecting that many changes; but it had to be done. The major changes are in PipArtifacts.cs and SandboxedProcessPipExecutor.cs files

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 1]
```diff
@@ -2233,24 +2240,55 @@ private bool PrepareInContainerPaths()
...
+                                var files = wrapper.Instance;
+                                FileUtilities.EnumerateDirectoryEntries(
+                                    directoryPathStr,
+                                    recursive: true,
+                                    handleEntry: (currentDir, name, attributes) =>
+                                    {
+                                        if ((attributes & FileAttributes.Directory) == 0)
+                                        {
+                                            var path = AbsolutePath.Create(m_pathTable, currentDir).Combine(m_pathTable, name);
+                                            var fileArtifact = FileArtifact.CreateOutputFile(path);
```
Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: It looks like the implementation of m_makeOutputPrivate immediate switches the FileArtifact to a string path. Can you just update that function signature and remove the translation?

File: Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs [hunk 2]
```diff
@@ -49,9 +49,45 @@ public static bool CanProduceOutputs(PipType pipType)
         /// <summary>
         /// Checks if a pip can preserved its outputs.
```
Comment (mpysson) [inline Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs]: >preserved [](start = 32, length = 9)

preserve

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs [hunk 3]
```diff
@@ -275,8 +310,65 @@ public void PreserveOutputsOpaqueTest()
...
             outputContents = RunSchedulerAndGetOutputContents(output, cacheHitAssert: false, id: pipA.Process.PipId);
             
-            // Preserve outputs is not supported for opaque dirs
+            // As the opaque output is preserved, the pip appended the existing file.
+            XAssert.AreEqual(CONTENT_TWICE, outputContents);
+
+            outputContents = RunSchedulerAndGetOutputContents(output, cacheHitAssert: true, id: pipA.Process.PipId);
+
+            // Cache hit and the appended file (CONTENT_TWICE) should remain the same.
+            XAssert.AreEqual(CONTENT_TWICE, outputContents);
```
Comment (mpysson) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs]: This is a breaking change. We should do a couple things
1. Make sure to include it in our release notes next week
2. Coordinate with Office to make sure they won't be blocked from taking the next release

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs [hunk 4]
```diff
@@ -519,7 +611,7 @@ public void PreservingProcessConsumingDynamicOutputTest()
             RunScheduler().AssertCacheMiss(preservingProcess.PipId, dynamicOutputProducer.PipId);
 
             // Delete dynamic output.
-            FileUtilities.DeleteDirectoryContents(ArtifactToString(outputDirectory), deleteRootDirectory: true);
+            //FileUtilities.DeleteDirectoryContents(ArtifactToString(outputDirectory), deleteRootDirectory: true);
```
Comment (mpysson) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs]: >//FileUtilities.DeleteDirectoryContents(ArtifactToString(outputDirectory), deleteRootDirectory: true); [](start = 12, length = 102)

It looks like this modifies the behavior of the test. Can you add it back in or did you changes cause some unintended consequences?

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/IntegrationTest.BuildXL.Scheduler.dsc [hunk 5]
```diff
@@ -15,7 +15,7 @@ namespace Scheduler.IntegrationTest {
         "AllowedUndeclaredReadsTests",
         "LazyMaterializationTests",
         "WhitelistTests",
-        "PreserveOutputsTests",
+        // "PreserveOutputsTests",
```
Comment (mpysson) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/IntegrationTest.BuildXL.Scheduler.dsc]: >//  [](start = 8, length = 3)

remove

File: Public/Src/Pips/Dll/Builders/ProcessBuilder.cs [hunk 6]
```diff
@@ -122,6 +122,9 @@ public class ProcessBuilder : IDisposable
         /// <nodoc />
         public Options Options { get; set; }
 
+        /// <nodoc />
+        public ReadOnlyArray<AbsolutePath> PreserveOutputWhitelist { get; set; } = ReadOnlyArray<AbsolutePath>.Empty;
```
Comment (mpysson) [inline Public/Src/Pips/Dll/Builders/ProcessBuilder.cs]: >PreserveOutputWhitelist [](start = 43, length = 23)

Do we need some error checking?
1. PreserveOutputWhitelist cannot be set unless PreserveOutputs is enabled
2. Members of PreserveOutputsWhitelist must be actual declared outputs and not stray paths or typos

File: Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs (hunk 2 above)
Comment (mpysson) [inline Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs]: >Checks if a pip can preserved its outputs. [](start = 12, length = 42)

Update comment to reflect this is now by file

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 7]
```diff
@@ -3736,5 +3773,28 @@ private void PreparePathForDirectory(string expandedDirectoryPath, bool createIf
                 FileUtilities.CreateDirectory(expandedDirectoryPath);
             }
         }
+
+        private bool ShouldPreserveOutput(AbsolutePath path, HashSet<AbsolutePath> whitelist)
```
Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >ShouldPreserveOutput [](start = 21, length = 20)

nit - this is only valid on declared output files and declared output directories. It would not be valid if called on a file contained within an output directory, unlike the method in PipArtifacts. I suggest adding some comments or modifying the function name to make this clear.

File: Public/Sdk/Public/Prelude/Prelude.OfficeBackCompatHacks.dsc [hunk 8]
```diff
@@ -151,6 +151,8 @@ namespace Transformer {
         allowPreservedOutputs?: boolean;
         /** Obsolete: You must use import {Transformer} from "Sdk.Transformers" */
         passThroughEnvironmentVariables?: string[];
+        /** Obsolete: You must use import {Transformer} from "Sdk.Transformers" */
+        preserveOutputWhitelist?: (File | Directory)[];
```
Comment (dannyvv) [inline Public/Sdk/Public/Prelude/Prelude.OfficeBackCompatHacks.dsc]: >preserveOutputWhitelist [](start = 8, length = 23)

Please do NOT add this one here...
TO use new API's office needs to modify the specgen, so they should at that time, add the import to the file.

This file is by now just here for the legacy branches which will not use that API.

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 9]
```diff
@@ -2116,19 +2116,26 @@ private void LogTempDirectoryTooLong(string directory)
...
-            }
-
-            if (!PrepareInContainerPaths())
-            {
-                return false;
-            }
-
+            using (var preserveOutputWhitelistWrapper = Pools.GetAbsolutePathSet())
             using (var dependenciesWrapper = Pools.GetAbsolutePathSet())
             using (var outputDirectoriesWrapper = Pools.GetAbsolutePathSet())
```
Comment (dannyvv) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: What are the odds of this function returning early in a succesfull build? If they are pretty high, maybe we should keep the pool access to the original location.
If low, this is a great change for readability.

File: Public/Sdk/Public/Transformers/Transformer.Execute.dsc [hunk 10]
```diff
@@ -219,6 +219,7 @@ namespace Transformer {
         hasUntrackedChildProcesses?: boolean;
         allowPreservedOutputs?: boolean;
         passThroughEnvironmentVariables?: string[];
+        preserveOutputWhitelist?: (File | Directory)[];
```
Comment (dannyvv) [inline Public/Sdk/Public/Transformers/Transformer.Execute.dsc]: Can you add a doc comment on what this does?
I had to read the code to infer the semantics as the name can imply both: "it should preserve output unless in this list list" or "it should purge unless it is in this list" as the term whitelist vs blacklist is dependent on the semantic :)

File: Public/Src/Engine/Scheduler/PipExecutor.cs [hunk 11]
```diff
@@ -4063,6 +4061,26 @@ private static bool IsProcessPreservingOutputs(IPipExecutionEnvironment environm
                    environment.Configuration.Sandbox.UnsafeSandboxConfiguration.PreserveOutputs != PreserveOutputsMode.Disabled;
         }
 
+        private static bool IsProcessPreservingOutputFile (IPipExecutionEnvironment environment, Process process, FileArtifact fileArtifact, FileOutputData fileOutputData)
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/PipExecutor.cs]: >  [](start = 57, length = 1)

nit: remove space

File: Public/Src/Pips/Dll/Operations/Process.cs [hunk 12]
```diff
@@ -596,6 +600,16 @@ public sealed partial class Process : Pip
...
 
+        /// <summary>
+        /// File/directory output paths that are preserved if <see cref="AllowPreserveOutputs"/> is enabled. 
+        /// </summary>
+        /// <remarks>
+        /// If the list is empty, all file and directory outputs are preserved. If the list is not empty,
+        /// only given paths are preserved and the rest is deleted.
+        /// </remarks>
+        [SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
+        public ReadOnlyArray<AbsolutePath> PreserveOutputWhitelist { get; }
```
Comment (narasamdya) [inline Public/Src/Pips/Dll/Operations/Process.cs]: >public ReadOnlyArray<AbsolutePath> PreserveOutputWhitelist { get; } [](start = 8, length = 67)

Should this affect fingerprint? 

Do we have unit test where we remove or add to the whitelist?

File: Public/Src/Pips/Dll/Builders/ProcessBuilder.cs (hunk 6 above)
Comment (narasamdya) [reply Public/Src/Pips/Dll/Builders/ProcessBuilder.cs]: This may be related to the fingerprint question that I wrote in some other comment.

Suppose that the whitelist affect weak fingerprint. Most likely We don't want any addition or removal of entries in the whitelist that have nothing to do with our outputs to affect our cacheability, right?

---
In reply to: [287428192](https://github.com/microsoft/BuildXL/pull/369#discussion_r287428192) [](ancestors = 287428192)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs (hunk 4 above)
Comment (narasamdya) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs]: CMIIW. I think the deletion of the outputDirectory there is meant to test file materialization.

---
In reply to: [287427160](https://github.com/microsoft/BuildXL/pull/369#discussion_r287427160) [](ancestors = 287427160)

File: Public/Src/Pips/Dll/Operations/Process.cs (hunk 12 above)
Comment (semihokur) [reply Public/Src/Pips/Dll/Operations/Process.cs]: Because “preserveOutputs” option does not affect the fingerprint, i thought that it makes sense that the whitelist also has no effect. 

I will add a test to modify the whitelist and expect a cache hit or miss depending on the availability of the output.

Comment (semihokur) [reply Public/Src/Pips/Dll/Operations/Process.cs]: Because “preserveOutputs” option does not affect the fingerprint, i thought that it makes sense that the whitelist also has no effect. 

I will add a test to modify the whitelist and expect a cache hit or miss depending on the availability of the output.

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 1 above)
Comment (semihokur) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: nice catch, it was pretty subtle. converted the parameter to string. 

---
In reply to: [287419128](https://github.com/microsoft/BuildXL/pull/369#discussion_r287419128) [](ancestors = 287419128)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs (hunk 3 above)
Comment (semihokur) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs]: I chatted with Ben this morning. He is aware of the change and he is fine with it. After we have the RC, he will get the changes done on the office side.

---
In reply to: [287426249](https://github.com/microsoft/BuildXL/pull/369#discussion_r287426249) [](ancestors = 287426249)

File: Public/Src/Pips/Dll/Builders/ProcessBuilder.cs (hunk 6 above)
Comment (semihokur) [reply Public/Src/Pips/Dll/Builders/ProcessBuilder.cs]: it is now affecting WF and added an unit test for that. 

I also added error-checking. 

---
In reply to: [287442996](https://github.com/microsoft/BuildXL/pull/369#discussion_r287442996) [](ancestors = 287442996,287428192)

File: Public/Sdk/Public/Transformers/Transformer.Execute.dsc (hunk 10 above)
Comment (semihokur) [reply Public/Sdk/Public/Transformers/Transformer.Execute.dsc]: agree, done!

---
In reply to: [287435027](https://github.com/microsoft/BuildXL/pull/369#discussion_r287435027) [](ancestors = 287435027)

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 9 above)
Comment (semihokur) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: early returns are all about failed builds, so we do not need to worry about it. 

---
In reply to: [287434293](https://github.com/microsoft/BuildXL/pull/369#discussion_r287434293) [](ancestors = 287434293)

File: Public/Sdk/Public/Prelude/Prelude.OfficeBackCompatHacks.dsc (hunk 8 above)
Comment (semihokur) [reply Public/Sdk/Public/Prelude/Prelude.OfficeBackCompatHacks.dsc]: gotcha! removed.

---
In reply to: [287433544](https://github.com/microsoft/BuildXL/pull/369#discussion_r287433544) [](ancestors = 287433544)

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 7 above)
Comment (semihokur) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: agree, it is confusing. 

renamed to ShouldPreserveDeclaredOutput and added comments. 

---
In reply to: [287430865](https://github.com/microsoft/BuildXL/pull/369#discussion_r287430865) [](ancestors = 287430865)

File: Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs (hunk 2 above)
Comment (semihokur) [reply Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs]: added remarks as well

---
In reply to: [287428462](https://github.com/microsoft/BuildXL/pull/369#discussion_r287428462) [](ancestors = 287428462)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs (hunk 4 above)
Comment (semihokur) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs]: I enabled that line again. But added the cachemiss assert after that: 



---
In reply to: [287444400](https://github.com/microsoft/BuildXL/pull/369#discussion_r287444400) [](ancestors = 287444400,287427160)

File: Public/Src/Engine/Scheduler/Graph/PipGraph.Builder.cs [hunk 13]
```diff
@@ -1304,6 +1308,24 @@ private bool IsValidIpc(IpcPip ipcPip, LockManager.PathAccessGroupLock pathAcces
                     return false;
                 }
 
+                if (process.PreserveOutputWhitelist.IsValid && process.PreserveOutputWhitelist.Length > 0)
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Graph/PipGraph.Builder.cs]: >if [](start = 16, length = 2)

I just want to ensure that we have unit test for this validation.

File: Public/Src/Engine/Scheduler/Tracing/Log.cs [hunk 14]
```diff
@@ -2596,6 +2596,38 @@ internal void FailedToHashInputFile(LoggingContext loggingContext, string pipDes
             string pipDescription,
             string pipValueId);
 
+        [GeneratedEvent(
+            (int)EventId.ScheduleFailAddPipDueToInvalidPreserveOutputWhitelist,
+            EventGenerators = EventGenerators.LocalOnly,
+            EventLevel = Level.Error,
+            Keywords = (int)(Events.Keywords.UserMessage | Events.Keywords.UserError),
+            EventTask = (int)Events.Tasks.Scheduler,
+            Message = "The pip '{pipDescription}' could not be added because one of PreserveOutputWhitelist is neither static file output nor directory output).")]
```
Comment (mpysson) [inline Public/Src/Engine/Scheduler/Tracing/Log.cs]: > could not be added because one of PreserveOutputWhitelist is neither static file output nor directory output [](start = 49, length = 109)

nit - suggest rewording to somethign like:
could not be added because it specified a PreserveOutputWhitelist which included {path} which was not defined as one of the pip's file or directory outputs.

File: Public/Src/Engine/Scheduler/Tracing/Log.cs [hunk 15]
```diff
@@ -2596,6 +2596,38 @@ internal void FailedToHashInputFile(LoggingContext loggingContext, string pipDes
...
+            string pipDescription,
+            string pipValueId);
+
+        [GeneratedEvent(
+            (int)EventId.ScheduleFailAddPipDueToInvalidAllowPreserveOutputsFlag,
+            EventGenerators = EventGenerators.LocalOnly,
+            EventLevel = Level.Error,
+            Keywords = (int)(Events.Keywords.UserMessage | Events.Keywords.UserError),
+            EventTask = (int)Events.Tasks.Scheduler,
+            Message = "The pip '{pipDescription}' could not be added because PreserveOutputWhitelist is set even though AllowPreserveOutputs is false for the pip).")]
```
Comment (mpysson) [inline Public/Src/Engine/Scheduler/Tracing/Log.cs]: >PreserveOutputWhitelist is set even though AllowPreserveOutputs is false for the pip [](start = 77, length = 84)

nit - PreserveOutputWhitelist requires AllowPreserveOutputs to be set to true.

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs (hunk 3 above)
Comment (mpysson) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/PreserveOutputsTests.cs]: I doubt it will matter to them but could double check on APEX and WDG as well?

---
In reply to: [287525535](https://github.com/microsoft/BuildXL/pull/369#discussion_r287525535) [](ancestors = 287525535,287426249)
